docs/reference/modules/discovery.asciidoc

```diff
@@ -1,30 +1,811 @@
-the main APIs (index, delete, search) do not communicate with the master
-node. The responsibility of the master node is to maintain the global
-cluster state, and act if nodes join or leave the cluster by reassigning
-shards. Each time a cluster state is changed, the state is made known to
-the other nodes in the cluster (the manner depends on the actual
-discovery implementation).
+It is separated into several sections, which are explained below:
+
+*   <<modules-discovery-hosts-providers,Discovery>> is the process where nodes
+    find each other when the master is unknown, such as when a node has just
+    started up or when the previous master has failed.
+*   <<modules-discovery-bootstrap-cluster>> is required when an Elasticsearch
+    cluster starts up for the very first time. In <<dev-vs-prod-mode,development
+    mode>>, with no discovery settings configured, this is automatically
+    performed by the nodes themselves. As this auto-bootstrapping is
+    <<modules-discovery-quorums,inherently unsafe>>, running a node in
+    <<dev-vs-prod-mode,production mode>> requires bootstrapping to be explicitly
+    configured via the `cluster.initial_master_nodes` setting.
+*   It is recommended to have a small and fixed number of master-eligible nodes
+    in a cluster, and to scale the cluster up and down by adding and removing
+    master-ineligible nodes only. However there are situations in which it may
+    be desirable to add or remove some master-eligible nodes to or from a
+    cluster. A section on <<modules-discovery-adding-removing-nodes,adding and
+    removing nodes>> describes this process as well as the extra steps that need
+    to be performed when removing more than half of the master-eligible nodes at
+    the same time.
+*   <<cluster-state-publishing,Cluster state publishing>> covers how a master
+    publishes cluster states to the other nodes in the cluster.
+*   The <<no-master-block,no-master block>> is put in place when there is no
+    known elected master, and can be configured to determine which operations
+    should be rejected when it is in place.
+*   <<master-election,Master election>> and <<fault-detection,fault detection>>
+    sections cover advanced settings to influence the election and fault
+    detection processes.
+*   <<modules-discovery-quorums,Quorum-based decision making>> explains the
+    design behind the master election and auto-reconfiguration logic.
+
+[float]
+[[modules-discovery-hosts-providers]]
+=== Discovery
+
+The cluster formation module uses a list of _seed_ nodes in order to start off
+the discovery process. At startup, or when disconnected from a master,
+Elasticsearch tries to connect to each seed node in its list, and holds a
+gossip-like conversation with them to find other nodes and to build a complete
+picture of the master-eligible nodes in the cluster. By default the cluster
+formation module offers two hosts providers to configure the list of seed
```

Comment:
```suggestion
formation module offers two host providers to configure the list of seed
```

Reply:
This occurs in other places too, so if this was not intentional, please fix all occurrence.

Reply:
Pluralising both "hosts" and "providers" is intentional: each provider provides many hosts, and there are many providers. I do see that "host providers" means pretty much the same thing, but this jargon is well-established. I think we are consistent in pluralising "hosts" everywhere (e.g. the setting is called `discovery.zen.hosts_provider`, the interface is `UnicastHostsProvider` and its implementations include `SettingsBasedHostsProvider` and `FileBasedUnicastHostsProvider`).